### DISCUSSION TITLE
Is it ok to mix rtti on/off c++ shared libraries

### DISCUSSION THREAD

**Discussion by skysky97** [android/ndk#1730] [Q&A]
I have an executable `app` compiled with `-frtti`, and a shared library `libnative.so` compiled with `-fno-rtti`.

STL: c++_shared
ndk: r21
The executable and library are built via ndk CMake.

In `app` code, throw `std::runtime_error` and try catch `std::exception`, it always failed.  

Example:

app.cpp

```c++
#include <exception>
#include <stdexcept>
#include <iostream>
#include <unistd.h>

int main() {
 try {
   throw std::runtime_error("User defined runtime error in app");
 }
 catch (const std::exception& e) {
 std::cout << "catch exception: " << e.what() << std::endl;
 }
 // catch (const std::runtime_error& e) {
 // std::cout << "catch runtime error: " << e.what() << std::endl;
 // }
 catch (...) {
 std::cout << "catch unknow exception" << std::endl;
 }
while(1) {
 sleep(5);
 }
}
```

libnative.cpp

```
#include <stdexcept>
void dummy() {
 throw std::logic_error("User defined logic error in libnative");
}
```

**Comment by DanAlbert** [comment]
> In app code, throw std::runtime_error and try catch std::exception, it always failed.

Define failed? I tried your code and everything is working as I expect. The `std::exception` catch handler catches the exception. Afaict the presence of libnative is also irrelevant. Here's my `CMakeLists.txt`:

```cmake
cmake_minimum_required(VERSION 3.23.1)
project(rtti-test CXX)

add_library(native SHARED native.cpp)
target_compile_options(native PRIVATE -fno-rtti)

add_executable(app app.cpp)
target_compile_options(app PRIVATE -frtti)
target_link_libraries(app PRIVATE native)
```

Built with

```bash
$ cmake -DCMAKE_TOOLCHAIN_FILE=~/Library/Android/sdk/ndk/24.0.8215888/build/cmake/android.toolchain.cmake -DANDROID_STL=c++_shared -DANDROID_ABI=arm64-v8a ..
$ make
```

`adb push`ed the outputs (and `libc++_shared.so`) to `/data/local/tmp` and ran with

```bash
adb shell "LD_LIBRARY_PATH=/data/local/tmp /data/local/tmp/app"
catch exception: User defined runtime error in app
```

**Reply by skysky97** [reply]
Failed means it always run into the `catch(...)` block.  

Yes, setting `LD_LIBRARY_PATH` for `libc++_shared.so` works fine. I'm sorry that I forgot to mention it.

But in my case, `libc++_shared.so` is located at `/system/lib64/vndk-sp-28/`, and `app` is located at `/vendor/bin/`, `libnative.so` is located at `/vendor/lib64/`. So `libc++_shared.so` is loaded into the `vendor` section `default` linker namespace (link to vndk namespace) by linker.  That makes test result different.

And when I throw `logic_error` exception and try catch, it runs into the `catch (std::exception)` block correctly. The difference is that `logic_error` appears in the libnative too.

**Reply by DanAlbert** [reply]
If it's a VNDK specific issue we're not going to be able to help. The only thing the NDK and VNDK have in common is the name. Best to follow b.android.com and file a bug with the appropriate section of the platform (or better yet, talk to your TAM).
